metis-media-service/src/main/java/eu/europeana/metis/mediaservice/MediaProcessor.java

```diff
@@ -0,0 +1,441 @@
+		command.addAll(Arrays.asList(
+				"-colorspace", "sRGB", "-dither", "Riemersma", "-remap", colormapFile.getPath(),
+				"-format", "\n%c", "histogram:info:"));
+		
+		List<String> results = runCommand(command, false);
+		
+		int width;
+		try {
+			width = Integer.parseInt(results.get(0));
+			int height = Integer.parseInt(results.get(1));
+			if (shouldExtractMetadata(urlTypes.get(url))) {
+				WebResource resource = edm.getWebResource(url);
+				resource.setMimeType(mimeType);
+				resource.setFileSize(content.length());
+				resource.setWidth(width);
+				resource.setHeight(height);
+				resource.setOrientation(width > height);
+				resource.setColorspace(results.get(2));
+				resource.setDominantColors(extractDominantColors(results));
+			}
+		} catch (Exception e) {
+			log.info("Could not parse ImageMagick response:\n" + StringUtils.join(results, "\n"), e);
+			throw new MediaException("File seems to be corrupted", "IMAGE ERROR");
+		}
+		
+		for (int i = 0; i < sizes; i++) {
+			File thumb = thumbs.get(i);
+			if (thumb.length() == 0)
+				throw new MediaException("Thumbnail file empty: " + thumb, "THUMBNAIL ERROR");
+			if (width < THUMB_SIZE[i]) {
+				FileUtils.copyFile(content, thumb);
+			}
+			thumbnails.put(thumb, url);
+		}
+	}
+	
+	/**
+	 * Thumbnails are put directly in temp dir with names that should be used in the
+	 * target storage. It's a problem when multiple EDMs share the same resource -
+	 * we can't use the same thumbnail files because system won't know when they can
+	 * be removed. So duplicate thumbnails are stored in subdirectories with
+	 * increasing numbers.
+	 */
+	private List<File> prepareThumbnailFiles(String url, String mimeType) throws IOException {
+		List<File> thumbs = new ArrayList<>();
+		String md5 = DigestUtils.md5Hex(url);
+		String ext = Arrays.asList("application/pdf", "image/png").contains(mimeType) ? ".png" : ".jpeg";
+		for (int i = 0; i < THUMB_SUFFIX.length; i++) {
+			File temp = File.createTempFile("thumb", null);
+			for (int j = 0; j < 1000; j++) {
+				File dir = j == 0 ? tempDir : new File(tempDir, "media_thumbnail_duplicates_" + j);
+				if (!dir.isDirectory() && !dir.mkdir())
+					throw new IOException("Could not create thumbnails subdirectory: " + dir);
+				try {
+					File f = new File(dir, md5 + THUMB_SUFFIX[i] + ext);
+					Files.move(temp.toPath(), f.toPath());
+					thumbs.add(f);
+					break;
+				} catch (FileAlreadyExistsException e) {
+					log.trace(j + " duplicates of " + url, e);
+					// try next dir
+				}
+			}
+			if (thumbs.size() < i + 1) {
+				if (!temp.delete())
+					log.warn("Could not remove temp file " + temp);
+				throw new IOException("Too many duplicates of url " + url);
+			}
+		}
+		return thumbs;
+	}
+	
+	private List<String> extractDominantColors(List<String> results) {
+		final Pattern pattern = Pattern.compile("#([0-9A-F]{6})");
+		return results.stream()
+				.skip(3)
```

Comment:
Constants for number 3 and 6. It's easier to understand if the reader know what these numbers refer to and why they are used.